common/environment/environment.cpp

```diff
@@ -1036,14 +1151,34 @@ void CLocalEnvironment::buildDropZoneCache() const
                 Owned<IConstEnvBase> cached = new CConstDropZoneInfo((CLocalEnvironment *) this, &it->query());
                 cache.setValue(x.str(), cached);
             }
-            name = it->query().queryProp("@computer");
-            if (name)
+            const char * computer = it->query().queryProp("@computer");
+            if (computer)
             {
+                StringBuffer xpath("@");
+                xpath.append(computer).append(NUMBER_OF_DROPZONES_SUFFIX);
+                unsigned numOfDropZoneByComputer = numOfDropzonesByComputer->getPropInt(xpath, -1);
+                if ( numOfDropZoneByComputer == -1 )
+                {
+                    numOfDropZoneByComputer = 0;
+                    numOfDropzonesByComputer->addPropInt(xpath, numOfDropZoneByComputer );
+                }
+                numOfDropZoneByComputer++;
+                numOfDropzonesByComputer->setPropInt(xpath, numOfDropZoneByComputer);
+
                 StringBuffer x("Software/DropZone[@computer=\"");
-                x.append(name).append("\"]");
+                x.append(computer);
+                x.append(DROPZONE_BY_MACHINE_SUFFIX).append(numOfDropZoneByComputer);
+                x.append("\"]");
+
                 Owned<IConstEnvBase> cached = new CConstDropZoneInfo((CLocalEnvironment *) this, &it->query());
                 cache.setValue(x.str(), cached);
             }
+
+            numOfDropZones++;
+            StringBuffer x("Software/DropZone[@id=\"");
+            x.append(DROPZONE_SUFFIX).append(numOfDropZones).append("\"]");
+            Owned<IConstEnvBase> cached = new CConstDropZoneInfo((CLocalEnvironment *) this, &it->query());
+            cache.setValue(x.str(), cached);
```

Comment:
I think this approach is ok, but it's working hard to keep using the existing string->IConstEnvBase cache map, for the purposes of later iterating. (See comment at top of diffs as well).

It would be cleaner to keep a simple array (or dedicated hashtable) of IConstDropZoneInfo, which can be iterated directly for by getDropZoneIterator().

As you have an additional 'numOfDropzonesByComputer' member anyway, you might as well have a IConstDropZoneInfo based list/table instead.